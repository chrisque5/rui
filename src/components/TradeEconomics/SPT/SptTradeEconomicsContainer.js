import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import moment from 'moment';
import TradeEconomics from '../TradeEconomics';
import SptTradeEconomics from './SptTradeEconomics';
import { getCurrencies } from '../../../utils/selectors';
import { actionTypes, ids } from '../../../utils/constants';
import { getSptDayCounts, resetDates } from '../../../actions/dateActions';
import { showWarningNotification, showErrorNotification } from '../../../utils/notifications';

const SptTradeEconomicsContainer = ({ form, setDateAndDayCount }) => {
  const dispatch = useDispatch();
  const currencies = useSelector(getCurrencies);

  const getInstrumentId = (currency1, currency2) => {
    const { instrumentId } = currencies.find((currency) => (
      currency.baseCurrency === currency1 && currency.counterCurrency === currency2)) || {};
    return instrumentId;
  };

  const currencyOnChange = (currency2) => {
    const currency1 = form.getFieldValue(ids.CURRENCY_1);
    if (currency1) {
      const instrumentId = getInstrumentId(currency1, currency2);
      if (instrumentId) { setDateAndDayCount(instrumentId); }
    }
  };

  const valueDateOnChange = async (valueDate) => {
    if (valueDate) {
      const currency1 = form.getFieldValue(ids.CURRENCY_1);
      const currency2 = form.getFieldValue(ids.CURRENCY_2);
      const dayDiff = moment().diff(valueDate, 'days');

      if (currency1 && currency2) {
        if ([0, 6].includes(valueDate.day()) || dayDiff > 0) {
          showWarningNotification('Date Warning', 'Value Date entered is a non-working day or in the past. Manually update all other dates.');
          form.resetFields([ids.RATE_1]);
        }

        const instrumentId = getInstrumentId(currency1, currency2);
        const tradeDate = form.getFieldValue(ids.TRADE_DATE) || null;

        const { dates, error = null, type } = await dispatch(getSptDayCounts(tradeDate, valueDate, instrumentId));

        if (type === actionTypes.LOAD_DATES_CANCELLED || type === actionTypes.LOAD_DATES_FAILED) {
          if (error) {
            dispatch(resetDates());
            showErrorNotification('Date Error', `${error.message || error.statusText}`);
          }
        } else {
          form.setFieldsValue({
            [ids.DAY_COUNT_1]: dates.dayCountFromTrade,
          });
        }
      }
    }
  };

  return (
    <SptTradeEconomics
      form={form}
      currencyOnChange={currencyOnChange}
      valueDateOnChange={valueDateOnChange}
    />
  );
};

SptTradeEconomicsContainer.propTypes = {
  form: PropTypes.shape().isRequired,
  setDateAndDayCount: PropTypes.func.isRequired,
};

export default TradeEconomics(SptTradeEconomicsContainer);
