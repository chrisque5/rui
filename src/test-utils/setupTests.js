// The below part of the code is to fix the issue
// TypeError: window.matchMedia is not a function
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
