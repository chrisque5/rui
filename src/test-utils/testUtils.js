/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Form } from 'antd';

const middlewares = [thunk];
const renderWithForm = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}><Form>{children}</Form></Provider>;

  return render(ui, { wrapper, ...renderOptions });
};

const renderWithProvider = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;

  return render(ui, { wrapper, ...renderOptions });
};
// re-export everything
export * from '@testing-library/react';
// override render method
export { renderWithForm, renderWithProvider };
