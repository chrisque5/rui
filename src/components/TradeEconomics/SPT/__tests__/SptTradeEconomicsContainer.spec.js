import { renderWithForm } from 'test-utils/testUtils';
import SptTradeEconomicsContainer from '../SptTradeEconomicsContainer';

jest.mock('../../TradeEconomics.js', () => (component) => component);
jest.mock('moment', () => () => jest.requireActual('moment')('2020-06-01T00:00:00.000Z'));
const mockForm = {
  getFieldValue: () => {},
  resetFields: () => {},
};

const initialState = {
  currencies: [],
  ui: { isInterestEnabled: false },
};

const defaultProps = {
  form: mockForm,
};

describe('<SptTradeEconomicsContainer />', () => {
  test('<SptTradeEconomicsContainer /> renders snapshot as expected', () => {
    const { container } = renderWithForm(<SptTradeEconomicsContainer {...defaultProps} />, { initialState });
    expect(container).toMatchSnapshot();
  });
});
