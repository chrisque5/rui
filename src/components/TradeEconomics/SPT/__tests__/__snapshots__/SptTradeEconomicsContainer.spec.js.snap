// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SptTradeEconomicsContainer /> <SptTradeEconomicsContainer /> renders snapshot as expected 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="economics-container"
    >
      <div
        class="item-row"
      >
        <div
          class="item-ccy"
        >
          <div
            class="ccy-pair"
          >
            <div
              class="ccy-label"
            >
              Currency Pair
              <span
                aria-label="star"
                class="anticon anticon-star star-icon"
                data-testid="addPreference"
                role="img"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  data-icon="star"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 00.6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0046.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
                  />
                </svg>
              </span>
              :
            </div>
            <div
              class="ant-row ant-form-item item-ccy1"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <div
                      class="ant-select ant-select-single ant-select-show-arrow ant-select-show-search"
                      data-testid="ddlCurrency1"
                      name="currency1"
                    >
                      <div
                        class="ant-select-selector"
                      >
                        <span
                          class="ant-select-selection-search"
                        >
                          <input
                            aria-activedescendant="currency1_list_0"
                            aria-autocomplete="list"
                            aria-controls="currency1_list"
                            aria-haspopup="listbox"
                            aria-owns="currency1_list"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            id="currency1"
                            role="combobox"
                            type="search"
                            value=""
                          />
                        </span>
                        <span
                          class="ant-select-selection-placeholder"
                        />
                      </div>
                      <span
                        aria-hidden="true"
                        class="ant-select-arrow"
                        style="user-select: none;"
                        unselectable="on"
                      >
                        <span
                          aria-label="down"
                          class="anticon anticon-down ant-select-suffix"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="down"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-row ant-form-item item-ccy2"
            >
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <div
                      class="ant-select ant-select-single ant-select-show-arrow ant-select-loading ant-select-show-search"
                      data-testid="ddlCurrency2"
                      name="currency2"
                    >
                      <div
                        class="ant-select-selector"
                      >
                        <span
                          class="ant-select-selection-search"
                        >
                          <input
                            aria-activedescendant="currency2_list_0"
                            aria-autocomplete="list"
                            aria-controls="currency2_list"
                            aria-haspopup="listbox"
                            aria-owns="currency2_list"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            id="currency2"
                            role="combobox"
                            type="search"
                            value=""
                          />
                        </span>
                        <span
                          class="ant-select-selection-placeholder"
                        />
                      </div>
                      <span
                        aria-hidden="true"
                        class="ant-select-arrow ant-select-arrow-loading"
                        style="user-select: none;"
                        unselectable="on"
                      >
                        <span
                          aria-label="loading"
                          class="anticon anticon-loading anticon-spin"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="loading"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="0 0 1024 1024"
                            width="1em"
                          >
                            <path
                              d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-row ant-form-item item-dealt-ccy"
          >
            <div
              class="ant-col ant-form-item-label"
            >
              <label
                class="ant-form-item-required"
                for="dealtCurrency"
                title="Dealt CCY"
              >
                Dealt CCY
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <div
                    class="ant-select ant-select-single ant-select-show-arrow ant-select-show-search"
                    data-testid="ddlCurrency3"
                    name="dealtCurrency"
                  >
                    <div
                      class="ant-select-selector"
                    >
                      <span
                        class="ant-select-selection-search"
                      >
                        <input
                          aria-activedescendant="dealtCurrency_list_0"
                          aria-autocomplete="list"
                          aria-controls="dealtCurrency_list"
                          aria-haspopup="listbox"
                          aria-owns="dealtCurrency_list"
                          autocomplete="off"
                          class="ant-select-selection-search-input"
                          id="dealtCurrency"
                          role="combobox"
                          type="search"
                          value=""
                        />
                      </span>
                      <span
                        class="ant-select-selection-placeholder"
                      />
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-arrow"
                      style="user-select: none;"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        class="anticon anticon-down ant-select-suffix"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rate-wrapper"
        >
          <div
            class="ant-row ant-form-item item-rate"
          >
            <div
              class="ant-col ant-form-item-label"
            >
              <label
                class=""
                for="rate1"
                title="Rate"
              >
                Rate
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <div
                    class="ant-input-number ant-input-number-sm"
                    style="width: 100%;"
                  >
                    <div
                      class="ant-input-number-handler-wrap"
                    >
                      <span
                        aria-disabled="false"
                        aria-label="Increase Value"
                        class="ant-input-number-handler ant-input-number-handler-up"
                        role="button"
                        unselectable="on"
                      >
                        <span
                          aria-label="up"
                          class="anticon anticon-up ant-input-number-handler-up-inner"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="up"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                      <span
                        aria-disabled="false"
                        aria-label="Decrease Value"
                        class="ant-input-number-handler ant-input-number-handler-down"
                        role="button"
                        unselectable="on"
                      >
                        <span
                          aria-label="down"
                          class="anticon anticon-down ant-input-number-handler-down-inner"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="down"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                    <div
                      class="ant-input-number-input-wrap"
                    >
                      <input
                        aria-valuemin="0"
                        autocomplete="off"
                        class="ant-input-number-input"
                        data-testid="txtPrice1"
                        id="rate1"
                        maxlength="15"
                        role="spinbutton"
                        step="1"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item item-amount"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="amount1"
              title="Amount"
            >
              Amount
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input ant-input-sm"
                  data-testid="txtAmount1"
                  id="amount1"
                  min="0"
                  style="width: 100%;"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item item-value-date"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="valueDate1"
              title=""
            >
              Value Date
              <span
                aria-label="star"
                class="anticon anticon-star star-icon"
                data-testid="addPreference"
                role="img"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  data-icon="star"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 00.6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0046.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
                  />
                </svg>
              </span>
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-picker"
                  style="width: 100%;"
                >
                  <div
                    class="ant-picker-input"
                  >
                    <input
                      autocomplete="off"
                      data-testid="dtpValueDate1"
                      id="valueDate1"
                      placeholder="Select date"
                      readonly=""
                      size="12"
                      tabindex="-1"
                      title="06/01/2020"
                      value="06/01/2020"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item item-day-count"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="dayCount1"
              title="Days"
            >
              Days
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-input-number ant-input-number-sm ant-input-number-disabled"
                  style="width: 100%;"
                >
                  <div
                    class="ant-input-number-handler-wrap"
                  >
                    <span
                      aria-disabled="false"
                      aria-label="Increase Value"
                      class="ant-input-number-handler ant-input-number-handler-up"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="up"
                        class="anticon anticon-up ant-input-number-handler-up-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="up"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                    <span
                      aria-disabled="false"
                      aria-label="Decrease Value"
                      class="ant-input-number-handler ant-input-number-handler-down"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        class="anticon anticon-down ant-input-number-handler-down-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                  <div
                    class="ant-input-number-input-wrap"
                  >
                    <input
                      aria-valuenow="0"
                      autocomplete="off"
                      class="ant-input-number-input"
                      data-testid="numDayCount1"
                      disabled=""
                      id="dayCount1"
                      role="spinbutton"
                      step="1"
                      value="0"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="item-cls"
        >
          <label
            class="label"
            for="cls1"
          >
            CLS: 
            <span
              aria-label="info-circle"
              class="anticon anticon-info-circle"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="info-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm32 664c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V456c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v272zm-32-344a48.01 48.01 0 010-96 48.01 48.01 0 010 96z"
                />
              </svg>
            </span>
          </label>
          <div
            class="ant-row ant-form-item cls"
          >
            <div
              class="ant-col ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <label
                    class="ant-checkbox-wrapper toggle"
                  >
                    <span
                      class="ant-checkbox"
                    >
                      <input
                        class="ant-checkbox-input"
                        data-testid="testCls1"
                        id="cls1"
                        type="checkbox"
                        value=""
                      />
                      <span
                        class="ant-checkbox-inner"
                      />
                    </span>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
